doc/src/guides/custom-functions.md

```diff
@@ -0,0 +1,1693 @@
+
+  Our example function $\operatorname{versin}(x)$ is a function from
+  $\mathbb{C}$ to $\mathbb{C}$, so it can accept any input. But suppose we had
+  a function that only made sense with certain inputs. As a second example,
+  let's define a function `divides` as
+
+  (custom-functions-divides-definition)=
+
+  $$\operatorname{divides}(m, n) = \begin{cases} 1 & \text{for}\: m \mid n \\
+  0 & \text{for}\: m\not\mid n  \end{cases}.$$
+
+  That is, `divides(m, n)` will be `1` if `m` divides `n` and `0` otherwise.
+  `divides` clearly only makes sense if `m` and `n` are integers.
+
+  We might be tempted to define the `eval()` method for `divides` like this:
+
+  ```py
+  >>> class divides(Function):
+  ...     @classmethod
+  ...     def eval(cls, m, n):
+  ...         # !! Not actually a good eval() method !!
+  ...
+  ...         # Evaluate for explicit integer m and n. This part is fine.
+  ...         if isinstance(m, Integer) and isinstance(n, Integer):
+  ...             return int(n % m == 0)
+  ...
+  ...         # For symbolic arguments, require m and n to be integer.
+  ...         # If we write the logic this way, we will run into trouble.
+  ...         if not m.is_integer or not n.is_integer:
+  ...             raise TypeError("m and n should be integers")
+  ```
+
+  The problem here is that by using `if not m.is_integer`, we are requiring
+  `m.is_integer` to be `True`. If it is `None`, it will fail (see the [guide
+  on booleans and three-valued logic](booleans-guide) for details on what it
+  means for an assumption to be `None`). This is problematic for two reasons.
+  Firstly, it forces the user to define assumptions on any input variable. If
+  the user omits them, it will fail:
+
+  ```py
+  >>> n, m = symbols('n m')
+  >>> print(n.is_integer)
+  None
+  >>> divides(m, n)
+  Traceback (most recent call last):
+  ...
+  TypeError: m and n should be integers
+  ```
+
+  Instead they have to write
+
+  ```py
+  >>> n, m = symbols('n m', integer=True)
+  >>> divides(m, n)
+  divides(m, n)
+  ```
+
+  This may seem like an acceptable restriction, but there is a bigger problem.
+  Sometimes, SymPy's assumptions system cannot deduce an assumption, even
+  though it is mathematically true. In this case, it will give `None` (`None`
+  means both "undefined" and "cannot compute" in SymPy's assumptions). For
+  example
+
+  ```py
+  >>> # n and m are still defined as integer=True as above
+  >>> divides(2, (m**2 + m)/2)
+  Traceback (most recent call last):
+  ...
+  TypeError: m and n should be integers
+  ```
+
+  Here the expression `(m**2 + m)/2` is always an integer, but SymPy's
+  assumptions system is not able to deduce this:
+
+  ```py
+  >>> print(((m**2 + m)/2).is_integer)
+  None
+  ```
+
+  SymPy's assumptions system is always improving, but there will always be
+  cases like this that it cannot deduce, due to the fundamental computational
+  complexity of the problem.
```

Comment:
The problem is not just computationally complex but is in fact mathematically impossible in general.

Reply:
I guess it's both. Sometimes a problem is decidable but intractable. And I'm hoping the assumptions system can evolve into a state where it only returns simple deductions by default and only tries harder when explicitly told to.